### PR TITLE
Re-introduce softmax with length in DotAttention (see #772)

### PR DESCRIPTION
**Pull request by fhieber**
Re-introduces #772 which was reverted after discovering an incompatibility with `broadcast_like` in MXNet 1.6. This is fixed in MXNet 1.7, therefore, this PR is based on the `mx17` branch.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/encoder.py [hunk 1]
```diff
@@ -328,11 +325,11 @@ def hybrid_forward(self, F, data, valid_length):
         if self.config.dropout_prepost > 0.0:
             data = F.Dropout(data=data, p=self.config.dropout_prepost)
 
-        # (batch_size * heads, 1, seq_len)
-        bias = F.expand_dims(self.valid_length_mask(data, valid_length), axis=1)
```
Comment (barbaradarques) [inline sockeye/encoder.py:332]: I like that the name is now consistent across classes. :)

File: sockeye/layers.py [hunk 2]
```diff
@@ -329,23 +304,17 @@ def hybrid_forward(self, F, queries, keys, values, lengths=None, bias=None):
...
-            # (n, lq, lk)
-            logits = F.transpose(data=logits, axes=(1, 2, 0))
-
         if bias is not None:
+            assert lengths is None, "Cannot provide both bias and lengths"
             logits = F.broadcast_add(logits, bias)
 
-        probs = F.softmax(logits, axis=-1)
+        if lengths is not None:
+            assert bias is None, "Cannot provide both bias and lengths"
```
Comment (barbaradarques) [inline sockeye/layers.py:312]: Wouldn't it it be better to assert this only once and in the beginning of the method? e.g.
```python
assert not (bias and lengths), "Cannot provide both bias and lengths"
```

Comment (fhieber) [reply sockeye/layers.py:312]: Yeah, that would be a bit cleaner. Actually, there is no reason to have this assert really. Of course it doesn't really make sense to use both 'modes' usually, but why not allow it?

**Review by barbaradarques — APPROVED** [review]
